domain/application/state/exposed.go

```diff
@@ -5,38 +5,481 @@ package state
+		}
+		if err := tx.Query(ctx, exposedSpacesStmt, ident).GetAll(&spaces); err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return errors.Errorf("retrieving exposed Spaces for application %q: %w", appID, err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	return encodeExposedEndopints(endpoints, cidrs, spaces), nil
+}
+
+func encodeExposedEndopints(endpoints []endpointName, cidrs []endpointCIDR, spaces []endpointSpace) map[string]application.ExposedEndpoint {
+	// We first need to init the map with a new empty set of strings for cidrs
+	// and spaces for each endpoint.
+	if len(endpoints) == 0 {
+		return nil
+	}
+	exposed := make(map[string]application.ExposedEndpoint, len(endpoints))
+	for _, endpoint := range endpoints {
+		endpointName := wildcardEndpointName
+		if endpoint.Name.Valid {
+			endpointName = endpoint.Name.String
+		}
+		exposed[endpointName] = application.ExposedEndpoint{
+			ExposeToCIDRs:    set.NewStrings(),
+			ExposeToSpaceIDs: set.NewStrings(),
+		}
+	}
+
+	for _, endpointCIDR := range cidrs {
+		endpointName := wildcardEndpointName
+		if endpointCIDR.EndpointName.Valid {
+			endpointName = endpointCIDR.EndpointName.String
+		}
+		exposed[endpointName].ExposeToCIDRs.Add(endpointCIDR.CIDR)
+	}
+	for _, endpointSpace := range spaces {
+		endpointName := wildcardEndpointName
+		if endpointSpace.EndpointName.Valid {
+			endpointName = endpointSpace.EndpointName.String
+		}
+		exposed[endpointName].ExposeToSpaceIDs.Add(endpointSpace.SpaceUUID)
+	}
+	return exposed
 }
 
 // UnsetExposeSettings removes the expose settings for the provided list of
 // endpoint names. If the resulting exposed endpoints map for the application
 // becomes empty after the settings are removed, the application will be
 // automatically unexposed.
-func (s *State) UnsetExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints set.Strings) error {
-	return nil
+func (st *State) UnsetExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints set.Strings) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for _, endpoint := range exposedEndpoints.Values() {
+			if err := st.unsetExposedEndpoint(ctx, tx, appID, endpoint); err != nil {
+				return errors.Capture(err)
+			}
+		}
+		return nil
+	})
+
+	return errors.Capture(err)
 }
 
 // MergeExposeSettings marks the application as exposed and merges the provided
 // ExposedEndpoint details into the current set of expose settings. The merge
 // operation will overwrite expose settings for each existing endpoint name.
-func (s *State) MergeExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints map[string]application.ExposedEndpoint) error {
+func (st *State) MergeExposeSettings(ctx context.Context, appID coreapplication.ID, exposedEndpoints map[string]application.ExposedEndpoint) error {
```

Comment:
This method (and the corresponding one in the service layer) should be renamed (probably to `SetExposedEndpoints()`) since the Merge verb is confusing: it was probably used because of the mongo document structure which required a merge of values in an array.
This will probably be done when wiring the apiserver.